docs(skill): add harness-authoring gotchas to add-harness-package (#16310)
vercel/ai: skills/add-harness-package/SKILL.md

Reviewer: unrelated to your changes, but this came to mind as I was reviewing:
```suggestion
- `"type": "module"`
- `"version": "0.0.0"` (starting point for new packages)
```

    └── pnpm-lock.yaml
```

> Add a `CHANGELOG.md` containing just the package heading (`# @ai-sdk/harness-<name>`). `konsistent` requires every package to have one, and the changeset release process fills in the rest — without it, `pnpm konsistent` fails.

Reviewer: maybe can shorten this to just:

```suggestion
Add a `CHANGELOG.md` containing just the package heading (`# @ai-sdk/harness-<name>`). Every package is required to have one.
```

the fact that `konsistent` fails without it isn't as important as it would later happen anyway

For bridge packages, add any bridge asset copy step required for files under `src/bridge/`.

Bridge dependency rules (bridge-backed harnesses):

> - The bridge's runtime deps live in `src/bridge/package.json` (installed in-sandbox at bootstrap), not the main package.json. After changing them, regenerate `src/bridge/pnpm-lock.yaml` with `pnpm install --lockfile-only --ignore-workspace` in that folder.

Reviewer: great point. FWIW I've been using:
```
pnpm --dir packages/<harness-package>/src/bridge install --lockfile-only --ignore-workspace
```

don't have to use it here, but it might be more obvious as it would allow running it from the repo root.

Author: addressed!

- For every third-party import in `src/bridge/`, keep three things in sync: the import, the `external` array in `tsup.config.ts`, and the dep in `src/bridge/package.json`. A missing entry shows up only at sandbox runtime as a module-resolution error.

> - Include packages the runtime _lazily_ imports — e.g. provider SDKs (`@langchain/anthropic`, `@langchain/openai`) resolved from the model id at runtime — even though nothing imports them directly. These fail only when a model of that provider is actually used.

Reviewer: nit: not ideal to have both examples be `@langchain/*` packages. maybe better to use independent examples, like Claude Code SDK, OpenAI SDK for instance.

Author: addressed!

- expose adapter-native built-in tools through `builtinTools`;
- keep construction synchronous and side-effect free;
- use `startOpts.sandboxSession` and `startOpts.sessionWorkDir`; never create a separate sandbox;
- throw `HarnessCapabilityUnsupportedError` from the method that needs an unsupported runtime capability;

> - don't hardcode a default model — pass the model only when the consumer configured one and let the runtime apply its own default, and keep the session's `modelId` consistent with what's actually sent (don't report a model the bridge silently overrode);

Reviewer: I would clarify that: don't hardcode a default model _unless_ it's technically required. Some of the underlying SDKs may not even have their own default.

this point is mostly about leaving the original SDK's default untouched, that's why we shouldn't pass one unless there is no original default.

- handle the `tools` and `instructions` that `doPromptTurn`/`doContinueTurn` may receive: if the runtime can't take custom `tools`, throw `HarnessCapabilityUnsupportedError` so it's obvious rather than silently dropped; if it has no native `instructions` input, prepend them to the first user message (the Codex/Claude Code workaround);
- quote interpolated paths (`workDir`, bridge-state dir, …) when building shell commands for `sandbox.run`/`sandbox.spawn` — they can contain spaces.

If the runtime needs in-sandbox setup, expose `getBootstrap()`.

### 5. Implement Runtime-Specific Concerns

Add only the concerns the runtime needs:

> - auth resolution — for AI Gateway support, use the central `getAiGatewayAuthFromEnv()` helper rather than reading env directly, and accept `VERCEL_OIDC_TOKEN` as a gateway credential (not just `AI_GATEWAY_API_KEY`). When the runtime resolves provider per model, resolve the provider from the model id and set that provider's env; if routing through the gateway, note that base-URL conventions differ per provider (e.g. an Anthropic client appends `/v1/messages` to a root base, an OpenAI client appends to a `/v1` base);

Reviewer: ```suggestion
- auth resolution — for AI Gateway support, use the central `getAiGatewayAuthFromEnv()` helper rather than reading env directly. This ensures both `VERCEL_OIDC_TOKEN` and `AI_GATEWAY_API_KEY` are accepted as Gateway credential. When the runtime resolves provider per model, resolve the provider from the model id and set that provider's env; if routing through the gateway, note that base-URL conventions differ per provider (e.g. an Anthropic client appends `/v1/messages` to a root base, an OpenAI client appends to a `/v1` base);
```
